diff --git a/bfd/config.bfd b/bfd/config.bfd
index bdee53957c0..04fff6e7df4 100644
--- a/bfd/config.bfd
+++ b/bfd/config.bfd
@@ -407,6 +407,12 @@ case "${targ}" in
     targ_selvecs="arm_pe_le_vec arm_pe_be_vec arm_pei_le_vec arm_pei_be_vec"
     targ_underscore=yes
     ;;
+  arm-*-mingw*)
+    targ_defvec=arm_pe_le_vec
+    targ_selvecs="arm_pe_le_vec arm_pe_be_vec arm_pei_le_vec arm_pei_be_vec"
+    targ_underscore=no
+    targ_cflags="-DARM_COFF_BUGFIX"
+    ;;
   arm-*-phoenix*)
     targ_defvec=arm_elf32_le_vec
     targ_selvecs=arm_elf32_be_vec
diff --git a/gas/configure.tgt b/gas/configure.tgt
index 3429f850d05..fa735da7ce6 100644
--- a/gas/configure.tgt
+++ b/gas/configure.tgt
@@ -164,7 +164,7 @@ case ${generic_target} in
   arm-*-nto*)				fmt=elf ;;
   arm-wince-pe | arm-*-wince | arm*-*-mingw32ce* | arm*-*-cegcc*)
 					fmt=coff em=wince-pe ;;
-  arm-*-pe)				fmt=coff em=pe ;;
+  arm-*-pe | arm-*-mingw*)	fmt=coff em=pe ;;
   arm-*-fuchsia*)			fmt=elf ;;
   arm-*-haiku*)				fmt=elf em=haiku ;;
 
